lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,292 @@
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	for _, mount := range mounts {
+		if mount.HostPath == "" || mount.ContainerPath == "" {
+			return nil, fmt.Errorf("The hostPath or containerPath is empty in the CDI mount: %v", *mount)
+		}
+
+		chosenOpts = []string{}
+		for _, opt := range mount.Options {
+			if !shared.ValueInSlice(opt, chosenOpts) {
+				chosenOpts = append(chosenOpts, opt)
+			}
+		}
+
+		chosenOptsStr := strings.Join(chosenOpts, ",")
+		splittedDevName = strings.Split(mount.HostPath, "/")
+		deviceName = splittedDevName[len(splittedDevName)-1]
+
+		// mount.HostPath can be a symbolic link, so we need to evaluate it
+		evaluatedHostPath, err := filepath.EvalSymlinks(mount.HostPath)
+		if err != nil {
+			return nil, err
+		}
+
+		if evaluatedHostPath != mount.HostPath && mount.ContainerPath == strings.TrimPrefix(mount.HostPath, rootPath) {
+			indirectSymlinks = append(indirectSymlinks, SymlinkEntry{Target: strings.TrimPrefix(evaluatedHostPath, rootPath), Link: mount.ContainerPath})
+			mount.ContainerPath = strings.TrimPrefix(evaluatedHostPath, rootPath)
+		}
+
+		configDevices.DiskDevs[deviceName] = map[string]string{
+			"type":              "disk",
+			"source":            evaluatedHostPath,
+			"path":              mount.ContainerPath,
+			"shift":             shift,
+			"raw.mount.options": chosenOptsStr,
+		}
+	}
+
+	// If the user desires to run a nested docker container inside a LXD container,
+	// the Tegra CSV files also need to be mounted so that the nvidia docker runtime
+	// can be auto-enabled as 'csv' mode.
+	if cdiID.Vendor == Nvidia && cdiID.Class == IGPU {
+		tegraCSVFilesCandidates := defaultNvidiaTegraCSVFiles(rootPath)
+		tegraCSVFiles := make([]string, 0)
+		for _, candidate := range tegraCSVFilesCandidates {
+			_, err := os.Stat(candidate)
+			if err == nil {
+				tegraCSVFiles = append(tegraCSVFiles, candidate)
+			} else if os.IsNotExist(err) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+
+		if len(tegraCSVFiles) == 0 {
+			return nil, fmt.Errorf("No CSV files detected for Tegra iGPU")
+		}
+
+		for _, tegraFile := range tegraCSVFiles {
+			splittedDevName = strings.Split(tegraFile, "/")
+			deviceName = splittedDevName[len(splittedDevName)-1]
+			configDevices.DiskDevs[deviceName] = map[string]string{
+				"type":     "disk",
+				"source":   tegraFile,
+				"path":     strings.TrimPrefix(tegraFile, rootPath),
+				"shift":    shift,
+				"readonly": "true",
+			}
+		}
+	}
+
+	return indirectSymlinks, nil
+}
+
+// specHookToLXDCDIHook will translate a hook from a CDI spec into an entry in a `Hooks`.
+// Some CDI hooks are not relevant for LXD and will be ignored.
+func specHookToLXDCDIHook(hook *specs.Hook, hooks *Hooks) error {
+	if hook == nil {
+		return fmt.Errorf("CDI hook is nil")
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	switch hook.Args[2] {
+	case "create-symlinks":
+		if len(hook.Args) < 5 {
```

Comment:
A general comment on these hooks. Checking the number of arguments like this seems a bit brittle and means that we may make changes that inadvertently break this functionality. There has been some discussion over including some hooks in the CDI spec -- at least from the perspective of their intent and they command line arguments.

With this in mind, we have recently introduced an additional exectuable: `nvidia-cdi-hook` that is specifically intended to implement these hooks. This means that `nvidia-ctk hook <HOOK_NAME>` and `nvidia-cdi-hook <HOOK_NAME>` are equivalent but would mean that your check here would fail for the latter since a minimal hook is:
```
nvidia-cdi-hook create-symlinks --link foo:bar
```
would fail this check since the length is just `4`.

My recommendation would be that we do something like the following:

```
	processHooks := map[string]func(string, []string) error {
		"create-symlinks":  processCreateSymlinksHook,
		"update-ldcache": processUpdateLdcacheHook,
	}

	for i, arg := range hook.Args {
		if process, supported := processHooks[arg]; supported {
			return process(rootPath, hook.Args[i:])
		}
	}

	return fmt.Errorf("Unsupported CDI hook function %q", hook.Args[2])
```

where 

```
func processCreateSymlinksHook(root string, args []string) error {
	return nil
}

func processUpdateLdcacheHook(root string, args []string) error {
	return nil
}
```
can be updated to implement the logic below.

Note that we may not want to return an error in this case and would just skip any unexpected hooks.

Reply:
Fixed